website/blog/deployers.md

```diff
@@ -0,0 +1,610 @@
+   run. A deployer should call `UpdateComponents` whenever there is a change to
+   the set of components a weavelet should be running.
+3. We call `envelope.Serve` to handle requests from the weavelet.
+
+```go
+// The unique id of the application deployment.
+var deploymentId = uuid.New().String()
+
+// spawn spawns the provided component (if it hasn't already spawned) and
+// returns the spawned weavelet hosting the component.
+func (d *deployer) spawn(component string) (*weavelet, error) {
+    d.mu.Lock()
+    defer d.mu.Unlock()
+
+    // Check if the component has already been spawned.
+    w, ok := d.weavelets[component]
+    if ok {
+        // The component has already been spawned.
+        return w, nil
+    }
+
+    // Spawn a weavelet to host the component.
+    info := &protos.EnvelopeInfo{
+        App:           "app",               // the application name
+        DeploymentId:  deploymentId,        // the deployment id
+        Id:            uuid.New().String(), // the weavelet id
+        SingleProcess: false,               // is the app a single process?
+        SingleMachine: true,                // is the app on a single machine?
+        RunMain:       component == "main", // should the weavelet run main?
+    }
+    config := &protos.AppConfig{
+        Name:   "app",       // the application name
+        Binary: flag.Arg(0), // the application binary
+    }
+    envelope, err := envelope.NewEnvelope(context.Background(), info, config)
+    if err != nil {
+        return nil, err
+    }
+
+    // Inform the weavelet of the component it should host.
+    if err := envelope.UpdateComponents([]string{component}); err != nil {
+        return nil, err
+    }
+
+    // Handle messages from the weavelet.
+    go envelope.Serve(&handler{d, envelope})
+
+    // Return the weavelet.
+    w = &weavelet{envelope: envelope, address: envelope.WeaveletInfo().DialAddr}
+    d.weavelets[component] = w
+    return w, nil
+}
+```
+
+Now, we implement the `EnvelopeHandler` methods.
+
+<div class="note">
+For each method, we also summarize the <a href="../docs.html#multiprocess"><code>weaver multi</code></a> and <a href="../docs.html#gke"><code>weaver gke</code><a/> implementations to give you a better sense for how these methods are implemented by more advanced deployers.
+</div>
+
+### Components
+
+First, we implement `ActiveComponent`. When a component hosted by a weavelet
+calls `weaver.Get[T]` for some component `T`, the `ActivateComponent` method
+gets invoked to activate `T`. `ActivateComponent` should start the
+component&mdash;potentially with multiple replicas&mdash;if it hasn't already
+been started.
+
+Our handler calls `deployer.spawn` to spawn the weavelet hosting the requested
+component, it it hasn't already been spawned. The handler then calls
+`UpdateRoutingInfo` to inform the weavelet of the requested component's address.
+This allows the weavelet to perform RPCs with the component.
```

Comment:
Here are we referring to the weavelet that's hosting this specific component? If so, the use of RPCs in weavelet -> component communication feels odd here, since the diagram at the beginning of the post made me believe that weavelet and components are linked into the same binary, so they'd perhaps communicate using method calls.